chain/ethevents/events.go

```diff
@@ -127,7 +133,108 @@ func NewEthEvents(contractsAddresses []common.Address, signer *ethereum.SignKeys
 			eventQueue:            make(map[string]*logEvent),
 			eventProcessorRunning: false,
 		},
-	}, nil
+	}
+
+	if scrutinizer != nil {
+		log.Infof("starting ethevents with scrutinizer enabled")
+		ethev.Scrutinizer = scrutinizer
+		ethev.Scrutinizer.AddEventListener(ethev)
+	}
+
+	return ethev, nil
+}
+
+// OnComputeResults is called once a process result is computed by the scrutinizer.
+func (ev *EthereumEvents) OnComputeResults(results *models.ProcessResult) {
+	tctx, cancel := context.WithTimeout(context.Background(), time.Minute*5)
+	defer cancel()
+	// check ethereum
+	// check ethereum process status
+	var fixedSizePID [types.ProcessIDsize]byte
+	copy(fixedSizePID[:], results.ProcessId)
+	ethProcessData, err := ev.VotingHandle.VotingProcess.Get(&ethbind.CallOpts{Context: tctx}, fixedSizePID)
+	if err != nil {
+		log.Errorf("error fetching process %x from Ethereum: %s", results.ProcessId, err)
+		return
+	}
+	if models.ProcessStatus(ethProcessData.Status) == models.ProcessStatus_RESULTS {
+		// check ethereum process results
+		ethResultsData, err := ev.VotingHandle.VotingProcess.GetResults(&ethbind.CallOpts{Context: tctx}, fixedSizePID)
+		if err != nil {
+			log.Errorf("error fetching process %x from Ethereum: %s", results.ProcessId, err)
+			return
+		}
+		if ethResultsData.Height > 0 {
+			log.Errorf("process %x results already added to ethereum", results.ProcessId)
+			return
+		}
+	} else if models.ProcessStatus(ethProcessData.Status) != models.ProcessStatus_ENDED {
+		log.Errorf("invalid process %x status for setting the results", results.ProcessId)
+		return
+	} else {
+		// valid eth
+		// check vochain
+		// check vochain process status
+		vocProcessData, err := ev.VochainApp.State.Process(results.ProcessId, true)
+		if err != nil {
+			log.Errorf("error fetching process %x from the Vochain: %s", results.ProcessId, err)
+			return
+		}
+		if models.ProcessStatus(vocProcessData.Status) == models.ProcessStatus_RESULTS {
+			// check vochain process results
+			if len(vocProcessData.Results.Votes) > 0 {
+				log.Errorf("process %x results already added on the Vochain", results.ProcessId)
+				return
+			}
+		} else if models.ProcessStatus(vocProcessData.Status) != models.ProcessStatus_ENDED {
+			log.Errorf("invalid process %x status %s for setting the results", results.ProcessId, vocProcessData.Status)
+			return
+		} else {
+			// check scrutinizer
+			_, err = ev.Scrutinizer.Storage.Get(ev.Scrutinizer.Encode("results", results.ProcessId))
+			if err == nil {
+				log.Errorf("process %x already computed", results.ProcessId)
+				return
```

Comment:
this func has a dozen returns preceded by log.Errorf - why not just return an error, then the caller can do the Errorf a single time?

Reply:
More or less what I commented before, the event listeners do not accept returning an error at this moment and adding so has some implications on the Vochain side. Happy to add it but needs some more analysis.

Reply:
TODO as before? :)